### ISSUE TITLE
Got StaleDataError when update_last_activity

### ISSUE THREAD

**Issue by Cas-pian** [jupyterhub/jupyterhub#2034]
**Describe the bug**
Sometimes I got error when update_last_activity runs function after login and start notebook for a while:`sqlalchemy.orm.exc.StaleDataError: UPDATE statement on table 'users' expected to update 1 row(s); 0 were matched.` 
**To Reproduce**
1. Start mysql as backend storage;
1. login and start notebook;
1. wait for a while (not sure how long) and then click the home and login out;

**Expected behavior**
No error of sqlachemy should be reported

**Desktop (please complete the following information):**
centos7.2.1511
chrome 67
0.8.1

**Additional context**
I'm using mysql cluster tool vitess.

**Comment by minrk** [comment]
Hm. I haven't seen that before. There have been some updates specifically to do with mysql compatibility in JupyterHub 0.9. Is it possible for you to upgrade and test the latest version? It would also be useful to get the complete traceback and some surrounding logs to see what operation is failing.

**Comment by Cas-pian** [comment]
Thanks for your reply!
I'm sorry I have to use 0.8.1 for the moment for some reasons. 
Could you please provide the compatibility fixes with mysql? 
The traceback is : 
```
[E 2018-07-16 09:57:18.043 JupyterHub app:1505] Rolling back session due to database error
    Traceback (most recent call last):
      File "/usr/lib/python3.4/site-packages/jupyterhub/app.py", line 1503, in update_last_activity
        self.db.commit()
      File "/usr/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 943, in commit
        self.transaction.commit()
      File "/usr/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 467, in commit
        self._prepare_impl()
      File "/usr/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 447, in _prepare_impl
        self.session.flush()
      File "/usr/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 2254, in flush
        self._flush(objects)
      File "/usr/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 2380, in _flush
        transaction.rollback(_capture_exception=True)
      File "/usr/lib/python3.4/site-packages/sqlalchemy/util/langhelpers.py", line 66, in __exit__
        compat.reraise(exc_type, exc_value, exc_tb)
      File "/usr/lib/python3.4/site-packages/sqlalchemy/util/compat.py", line 187, in reraise
        raise value
      File "/usr/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 2344, in _flush
        flush_context.execute()
      File "/usr/lib/python3.4/site-packages/sqlalchemy/orm/unitofwork.py", line 391, in execute
        rec.execute(self)
      File "/usr/lib/python3.4/site-packages/sqlalchemy/orm/unitofwork.py", line 556, in execute
        uow
      File "/usr/lib/python3.4/site-packages/sqlalchemy/orm/persistence.py", line 177, in save_obj
        mapper, table, update)
      File "/usr/lib/python3.4/site-packages/sqlalchemy/orm/persistence.py", line 791, in _emit_update_statements
        (table.description, len(records), rows))
    sqlalchemy.orm.exc.StaleDataError: UPDATE statement on table 'users' expected to update 1 row(s); 0 were matched.
```

**Comment by sougou** [comment]
I happened to stumble upon this (I work on vitess). You may be hitting an issue related to CLIENT_FOUND_ROWS setting. Vitess connection pools turn this off by default. You can make vitess use a special pool with that flag turned on by issuing a `set client_found_rows = 1` before starting your transactions.

**Comment by Cas-pian** [comment]
@sougou Thanks for you reply. I'll try it.

**Comment by Cas-pian** [comment]
@sougou According to [this](http://docs.sqlalchemy.org/en/latest/dialects/mysql.html?highlight=found_rows#rowcount-support)， it's set by default. So any other suggestions?

**Comment by sougou** [comment]
Did you also call `set client_found_rows=1` after obtaining the connection?

**Comment by Cas-pian** [comment]
Yes, I've already done that. And I'm not sure whether pymysql is compatible with vitess? 
Im' using python3, sqlalchemy, pymysql to connect vitess backend.
Any suggestion is appreciated, Thanks!

**Comment by sougou** [comment]
You can try this:
* Start a transaction
* Perform an update that actually changes a row (a value in the row needs to change). This should not fail.
* Before closing the transaction, check the vttablet URL `/debug/vars`. There should be two json entries: `TransactionPool` and `FoundRowsPool`. If you set the variable correctly, the one connection should be active in `FoundRowsPool`. If the value didn't take, then one will be active in `TransactionPool`.

Note that the default transaction timeout is 30s. You'll have to do the check before it times out, or increase the timeout to give yourself enough time.

**Comment by manics** [comment]
I'm closing this as it's quite old and there's no concrete action point. If you're still having problems please post on the [Jupyter Community Forum](https://discourse.jupyter.org/) or open a new issue. Thanks!
